### DISCUSSION TITLE
Display updated value from API call

### DISCUSSION THREAD

**Discussion by cybardev** [alpinejs/alpine#3611] [1. Help] [answered]
## Code

`CodePen`: [`cybarspace/BaGBONG`](https://codepen.io/cybarspace/pen/BaGBONG)

<details><summary><b>View full code...</b></summary><br />

`HTML` (`<body>` contents)
```html
<section x-data="gpsData">
  <aside x-init="getGPSData">
    <p>
      <span>Num. of Satellites</span>:&nbsp;<span x-text="numSatellites"></span>
    </p>
    <p>
      <span>Latitude</span>:&nbsp;<span x-text="latitude"></span>
    </p>
    <p>
      <span>Longitude</span>:&nbsp;<span x-text="longitude"></span>
    </p>
  </aside>
</section>
<script defer src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js"></script>
```

`JavaScript`
```js
const gpsData = new (class {
  // attributes
  numSatellites = 0;
  latitude = 0;
  longitude = 0;

  // methods
  getGPSData = () => {
    this.numSatellites = fetch(API_URLS.numSatellites)
      .then(UTILS.res)
      .catch(UTILS.err);
    this.latitude = fetch(API_URLS.latitude)
      .then(UTILS.res)
      .catch(UTILS.err);
    this.longitude = fetch(API_URLS.longitude)
      .then(UTILS.res)
      .catch(UTILS.err);

    // update GPS data properties
    setTimeout(() => this.getGPSData(), 1000);
  };
})();

const API_URLS = {
  // GPS endpoints
  numSatellites:
    "https://demo.signalk.org/signalk/v1/api/vessels/self/navigation/gnss/satellites/value/",
  latitude:
    "https://demo.signalk.org/signalk/v1/api/vessels/self/navigation/position/value/latitude/",
  longitude:
    "https://demo.signalk.org/signalk/v1/api/vessels/self/navigation/position/value/longitude/"
};

const UTILS = {
  res(msg) {
    return msg.text();
  },
  err(msg) {
    return console.log(msg);
  }
};
```

`CSS`
```css
@import url("https://unpkg.com/mvp.css@1.12/mvp.css");

aside > p > span:nth-child(1) {
  font-weight: bold;
}
```

</details>

## Goal

- [x] `x-init` fetches initial data
- [x] `x-text` populates initial data
- [x] `setTimeout` callback fetches data periodically
- [x] after retrieving data, it's updated in `x-text`

## Problem

This is an extension of #3608. The difference with that is that this uses `fetch` calls to retrieve data from an API endpoint.

I've read the [Async](https://alpinejs.dev/advanced/async) page of the Alpine docs. Here's what I tried from reading that, but it didn't work (check **Screenshots** below):

```js
getGPSData = async () => {
  this.numSatellites = await fetch(API_URLS.numSatellites)
    .then(UTILS.res)
    .catch(UTILS.err);
  this.latitude = await fetch(API_URLS.latitude)
    .then(UTILS.res)
    .catch(UTILS.err);
  this.longitude = await fetch(API_URLS.longitude)
    .then(UTILS.res)
    .catch(UTILS.err);

  // update GPS data properties
  setTimeout(() => this.getGPSData(), 1000);
};
```

I'm certainly doing something wrong, but I'm lost on what that might be.

**PS**: Interestingly, the values are updated if you check in the console either way.

## Screenshots

<figure>
<img width="355" alt="screenshot of before trying async/await" src="https://github.com/alpinejs/alpine/assets/50134239/e0955060-5061-4269-809c-3037a92c833a" title="screenshot of before trying async/await">
<br />
<figcaption>Before trying <code>async</code>/<code>await</code></figcaption>
</figure>

<hr />

<figure>
<img width="353" alt="screenshot of after trying async/await" src="https://github.com/alpinejs/alpine/assets/50134239/7a0fc8fb-676c-4dbc-b5e1-f38b8d213f11" title="screenshot of after trying async/await">
<br />
<figcaption>After trying <code>async</code>/<code>await</code></figcaption>
</figure>

## Request

Any help regarding "how to update `x-text` when object property changes after `fetch` call" would be much appreciated.

**Comment by ekwoka** [comment, marked as answer]
You need to not use a class like that, and not use arrow functions for methods.

The choice to use a class like that is very confusing...

But you need to ensure your methods are acting on the reactive context with their `this` not on the underlying object.

instead of 

```js
const gpsData = new (class {
  // attributes
  numSatellites = 0;
  latitude = 0;
  longitude = 0;

  // methods
  getGPSData = () => {
    this.numSatellites = fetch(API_URLS.numSatellites)
      .then(UTILS.res)
      .catch(UTILS.err);
    this.latitude = fetch(API_URLS.latitude)
      .then(UTILS.res)
      .catch(UTILS.err);
    this.longitude = fetch(API_URLS.longitude)
      .then(UTILS.res)
      .catch(UTILS.err);

    // update GPS data properties
    setTimeout(() => this.getGPSData(), 1000);
  };
})();
```

just do

```js
const gpsData = () => ({
  numSatellites: 0,
  latitude: 0,
  longitude: 0,
  init() {
    setInterval(() => this.getGPSData(), 1000)
  },
  async getGPSData() {
    [this.numSatellites, this.latitude, this.longitude] = await Promise.all([
      fetchText(API_URLS.numSatellites),
      fetchText(API_URLS.latitude),
      fetchText(API_URL.longitude)
    ])
  }
})

const fetchText = (url) => fetch(url).then(res => res.text())
```

**Reply by cybardev** [reply]
That makes sense. I think without using a λ class, `this` was in the global scope instead of local.

The λ function looks more elegant tho and I'll try that tomorrow when I'm in office.

Thanks for the help.

**Reply by ekwoka** [reply]
Factory functions is the appropriate way to use components in Alpine

**Reply by cybardev** [reply]
Tested to work. Thank you very much for your kind and patient help again 😊 Learning a lot.

**Reply by cybardev** [reply]
Loosely related. Any suggestions on how to improve this in JS? It looks kinda hacky...

```js
const API_URLS = {
    // origin
    server: "https://demo.signalk.org/signalk/v1/api/",

    // sublevels
    get gps() {
        return this.server + "vessels/self/navigation/";
    },

    // GPS endpoints
    get numSatellites() {
        return this.gps + "gnss/satellites/value/";
    },
    get latitude() {
        return this.gps + "position/value/latitude/";
    },
    get longitude() {
        return this.gps + "position/value/longitude/";
    },
}
```

I need this cuz I have more sensors other than just GPS at the same `server` API, and it would be bad if I had to repeat the same huge string every time. Plus it wouldn't be easily modifiable if the server address changed.

---

**UPDATE**

Did it like this:

```js
const API_URLS = {
    // origin
    server: "https://demo.signalk.org/signalk/v1/api/",

    init() {
        // sublevels
        this.gps = this.server + "vessels/self/navigation/";

        // GPS endpoints
        this.numSatellites = this.gps + "gnss/satellites/value/";
        this.latitude = this.gps + "position/value/latitude/";
        this.longitude = this.gps + "position/value/longitude/";

        // TODO: other attributes here...

        // initialize self
        delete this.init;
        return this;
    },
}.init();
```

It's more hacky, but looks nicer.

**Reply by ekwoka** [reply]
Honestly, that's worse.

If you want to programmatically pre-generate them, don't use an object and method.

Just generate them

```js
const base = 'blah.com/'

export const API_URLS = {
  satellites: base + 'satallites'
}
```

**Reply by cybardev** [reply]
That looks good. Thanks a lot. 😊

**Reply by cybardev** [reply]
Wait, that breaks when I try to add more stuff, like:

```js
const base = "blah.com/"

export const API_URLS = {
  satellites: base + "satallites/",

  orion: this.satellites + "orion/"
  artemis: this.satellites + "artemis/"
  altair: this.satellites + "altair/"
}
```

I just don't want to pollute the global scope with a bunch of constants, so I was thinking of encapsulating them in an object. If there's no other way then maybe I should just stick to my λ class way (for just the URLs):

```js
const API_URLS = new (class {
    base = "blah.com/";
    satellites = this.base + "satallites/";

    orion = this.satellites + "orion/";
    artemis = this.satellites + "artemis/";
    altair = this.satellites + "altair/";
})();
```

**Reply by ekwoka** [reply]
You cant use `this` like that. It would point to the `globalThis` not the object.

Stop using IICE, those aren't a thing.

It's not polluting the global scope if you use modules correctly.

I'm honestly so extremely confused by how you got to doing this instantly invoked class expression. It's the wildest thing I've seen all year.

You make a file like `urls.js`

you do

```js
const base = 'blah.com/'

const satellites_base = base + "satallites/"

export const API_URLS = {
  satellites: {
    orion: satellites_base + 'orion'
  }
}
```

**Reply by cybardev** [reply]
No idea what `IICE` is...

Also, `orion` isn't the end of it. It might have further sublevels, like:

```js
const API_URLS = new (class {
    base = "blah.com/";
    satellites = this.base + "satallites/";

    orion = this.satellites + "orion/";
    artemis = this.satellites + "artemis/";
    altair = this.satellites + "altair/";

    room1 = this.orion + "room1/"
    room2 = this.orion + "room2/"

    cubicle1 = this.room1 + "cubicle1/"
    cubicle2 = this.room1 + "cubicle2/"
})();
```

I'll look into making it multi-file/modular, thanks for the suggestion. Currently I'm just using a single `script.js` (client-side).

The application is a Dashboard that gets info from an API and displays them in blocks.

---

> I'm honestly so extremely confused by how you got to doing this instantly invoked class expression.

That's cuz of...

> You cant use `this` like that. It would point to the `globalThis` not the object.

**Reply by ekwoka** [reply]
Immediately invoked class expression. You've never heard of it because there is no reason to ever do it.

Regardless of number of levels, doesn't make sense to do an IICE. Even an IIFE would be better and take less syntax.

**Reply by cybardev** [reply]
Tried `IIFE` but the concatenation build-up doesn't work (cuz I'm obviously doing it very wrong, but I don't know what the right way is). Example:

```js
// IIFE doesn't work
const F = (() => ({
    a: "A",
    b: this.a + "B",
    c: this.b + "C",
    d: this.a + "D",
}))();

// IICE does work
const C = new (class {
    a = "A";
    b = this.a + "B";
    c = this.b + "C";
    d = this.a + "D";
})();
```

I *need* things to **accumulate** on top of one of the last things. If `this` isn't the way to do it in `IIFE` then *what is*?

From looking at [this StackOverflow thread](https://stackoverflow.com/q/4616202), I don't see anything that might fit my needs stated above. Please do correct me if I am wrong (which is very likely cuz again, I lack knowledge).

**Reply by ekwoka** [reply]
Please stop trying to use `this`... It's just more characters for no benefit....

Just use a variable...

```js
const F = (() => {
  const a = 'A'
  return {
    b: a + "B",
    c: b + "C",
    d: a + "D",
  }
})();
```


But really, at this point it almost seems like you want to do more like

```js
const segments = {
  A: [
    'B'
  ]
}
```

And then building it from the tree.

**Reply by cybardev** [reply]
```js
const F = (() => {
  const a = 'A'
  return {
    b: a + "B",
    c: b + "C",
    d: a + "D",
  }
})();
```
Gives `Uncaught ReferenceError: b is not defined` where you do `c: b + "C",`

**Reply by cybardev** [reply]
Ok, here's the full context:

```js
const API_URLS = new (class {
    // host
    host = "demo.signalk.org"; // "localhost:3443",

    // origin
    server = "https://" + this.host + "/signalk/v1/api/";
    self = this.server + "vessels/self/";

    // sublevels
    gps = this.self + "navigation/";
    ais = this.server + "vessels/";
    depth = this.self + "environment/depth/";
    audio = this.self + "audio/";

    // GPS endpoints
    numSatellites = this.gps + "gnss/satellites/value/";
    latitude = this.gps + "position/value/latitude/";
    longitude = this.gps + "position/value/longitude/";

    // Depth endpoints
    belowTransducer = this.depth + "belowTransducer/value/";
    transducerToKeel = this.depth + "transducerToKeel/value/";
    belowKeel = this.depth + "belowKeel/value/";

    // Audio endpoints
    file = this.audio + "file/";
    status = this.audio + "status/";
})();
```

This works. You're saying that there's a better way. But I either don't know the better ways, or can't get them to work because of my lack of knowledge.

**Reply by ekwoka** [reply]
```js
// origin
const HOST = "demo.signalk.org"; // "localhost:3443",
const SERVER = `https://${HOST}/signalk/v1/api/`;
const SELF = SERVER + "vessels/self/";

// sublevels
const GPS = SELF + "navigation/";
const AIS = SERVER + "vessels/";
const DEPTH = SELF + "environment/depth/";
const AUDIO = SELF + "audio/";

export const API_ENDPOINTS = {
  // GPS endpoints
  numSatellites: GPS + "gnss/satellites/value/",
  latitude: GPS + "position/value/latitude/",
  longitude: GPS + "position/value/longitude/",

  // Depth endpoints
  belowTransducer: DEPTH + "belowTransducer/value/",
  transducerToKeel: DEPTH + "transducerToKeel/value/",
  belowKeel: DEPTH + "belowKeel/value/",

  // Audio endpoints
  file: AUDIO + "file/",
  status: AUDIO + "status/"
}
```

This is the typical way to do this.

**Reply by cybardev** [reply]
Thanks for the suggestion, but I couldn't get it to work. I'll keep trying in my time off, but a bit before I read your comment, I decided to rewrite the components entirely from scratch with a different structure so I won't need a big object like that just to encapsulate URLs. I'll put the `origin` section in the global scope and make components like so:

`script.js`
```js
// origin
const API = "https://localhost:3443/signalk/v1/api/";

// component
const DashboardComponent = () => ({
    meta: {
        title: "Component",
        data: ["x", "y", "z"],
        endpoint: API + "data/",
        show: true,
    },

    data: [0, 0, 0],

    async init() {
        [this.data[0], this.data[1], this.data[2]] = await Promise.all([
            UTILS.fetchText(this.meta.endpoint + "x/"),
            UTILS.fetchText(this.meta.endpoint + "y/"),
            UTILS.fetchText(this.meta.endpoint + "z/"),
        ]);

        // refresh data every 3 seconds
        setTimeout(() => this.init(), 3000);
    },
});
```

`index.html`
```html
<aside x-cloak x-data="DashboardComponent" x-show="meta.show" x-transition>
    <h3 x-text="meta.title"></h3>
    <hr />
    <p>
        <span x-text="meta.data[0]"></span>:
        <span x-text="data[0]"></span>
    </p>
    <p>
        <span x-text="meta.data[1]"></span>:
        <span x-text="data[1]"></span>
    </p>
</aside>
```
